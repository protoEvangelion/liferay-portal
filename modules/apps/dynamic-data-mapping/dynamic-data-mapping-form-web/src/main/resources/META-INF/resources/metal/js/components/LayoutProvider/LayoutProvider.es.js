import {Config} from 'metal-state';
import {FormSupport} from '../Form/index.es';
import {pageStructure} from '../../util/config.es';
import {sub} from '../../util/strings.es';
import Component from 'metal-jsx';

/**
 * LayoutProvider listens to your children's events to
 * control the `pages` and make manipulations.
 * @extends Component
 */

class LayoutProvider extends Component {
	static PROPS = {

		/**
		 * @default undefined
		 * @instance
		 * @memberof LayoutProvider
		 * @type {?(array|undefined)}
		 */

		pages: Config.arrayOf(pageStructure).value([]),

		/**
		 * @default undefined
		 * @instance
		 * @memberof LayoutProvider
		 * @type {?(array|undefined)}
		 */

		spritemap: Config.string()
	};

	static STATE = {

		/**
		 * @instance
		 * @memberof FormPage
		 * @type {?number}
		 */

		activePage: Config.number().value(0),

		/**
		 * @default undefined
		 * @instance
		 * @memberof LayoutProvider
		 * @type {?array}
		 */

		pages: Config.array(),

		/**
		 * @default undefined
		 * @instance
		 * @memberof LayoutProvider
		 * @type {?object}
		 */

		focusedField: Config.shapeOf(
			{
				columnIndex: Config.oneOfType(
					[
						Config.bool().value(false),
						Config.number()
					]
				).required(),
				pageIndex: Config.number().required(),
				rowIndex: Config.number().required(),
				type: Config.string().required()
			}
		).value({}),

		/**
		 * @default add
		 * @instance
		 * @memberof LayoutProvider
		 * @type {?string}
		 */

		mode: Config.oneOf(['add', 'edit']).value('add')
	};

	/**
	 * @inheritDoc
	 */

	constructor(props, pages) {
		super(props, pages);

		this.state.pages = props.pages;
	}

	/**
	 * @param {!Object} data
	 * @private
	 */

	_handleClickedField(data) {
		this.setState(
			{
				focusedField: data,
				mode: 'edit'
			}
		);
	}

	/**
	 * @param {!Object} payload
	 * @private
	 */

	_handleFieldAdd({target, fieldProperties}) {
		const {spritemap} = this.props;
		const {pages} = this.state;
		const {columnIndex, pageIndex, rowIndex} = target;

		fieldProperties = Object.assign({}, fieldProperties, {spritemap});

		let newContext = null;

		if (target.columnIndex === false) {
			const newRow = FormSupport.implAddRow(12, [fieldProperties]);

			newContext = FormSupport.addRow(
				pages,
				rowIndex,
				pageIndex,
				newRow
			);
		}
		else {
			newContext = FormSupport.addFieldToColumn(
				pages,
				pageIndex,
				rowIndex,
				columnIndex,
				fieldProperties
			);
		}

		this.setState(
			{
				focusedField: {
					columnIndex,
					pageIndex,
					rowIndex,
					type: fieldProperties.type
				},
				mode: 'edit',
				pages: newContext
			}
		);
	}

	/**
	 * @param {!Object} event
	 * @private
	 */

	_handleDeleteField({rowIndex, pageIndex, columnIndex}) {
		const {pages} = this.state;
		let newContext = FormSupport.removeFields(
			pages,
			pageIndex,
			rowIndex,
			columnIndex
		);

		newContext = this._removeEmptyRow(
			newContext,
			{
				columnIndex,
				pageIndex,
				rowIndex
			}
		);

		this.setState(
			{
				focusedField: {},
				pages: newContext
			}
		);
	}

	/**
	 * @param {!Object}
	 * @private
	 */

	_handleDuplicatedField({rowIndex, pageIndex, columnIndex}) {
		const {pages} = this.state;
		const field = FormSupport.getField(pages, pageIndex, rowIndex, columnIndex);

		const duplicatedField = {
			...field,
			label: sub(
				Liferay.Language.get('copy-of-x'),
				[field.label]
			),
			name: FormSupport.generateFieldName(field)
		};

		const newRowIndex = rowIndex + 1;

		const newContext = FormSupport.addRow(pages, newRowIndex, pageIndex);

		FormSupport.addFieldToColumn(newContext, pageIndex, newRowIndex, columnIndex, duplicatedField);

		this.setState(
			{
				pages: newContext
			}
		);
	}

	/**
	 * @param {!Object} event
	 * @private
	 */

	_handleFieldEdited({value, key}) {
		const {focusedField, pages} = this.state;
		const {columnIndex, pageIndex, rowIndex} = focusedField;
		const column = FormSupport.getColumn(
			pages,
			pageIndex,
			rowIndex,
			columnIndex
		);

		const implPropertiesField = {
			[key]: value
		};

		const newField = Object.assign(
			{},
			column.fields[0],
			implPropertiesField
		);

		FormSupport.changeFieldsFromColumn(
			pages,
			pageIndex,
			rowIndex,
			columnIndex,
			[newField]
		);

		this.setState(
			{
				pages: this.state.pages
			}
		);
	}

	/**
	 * @param {!Object} event
	 * @private
	 */

	_handleFieldMoved({target, source}) {
		const {pages} = this.state;
		const {columnIndex, pageIndex, rowIndex} = source;
		const column = FormSupport.getColumn(
			pages,
			pageIndex,
			rowIndex,
			columnIndex
		);
		const {fields} = column;

		let newContext = FormSupport.removeFields(
			pages,
			pageIndex,
			rowIndex,
			columnIndex
		);

		if (target.columnIndex === false) {
			newContext = this._removeEmptyRow(newContext, source);
			newContext = this._addRow(
				newContext,
				target,
				fields
			);
		}
		else {
			newContext = this._setColumnFields(
				newContext,
				target,
				fields
			);
			newContext = this._removeEmptyRow(newContext, source);
		}

		this.setState(
			{
				focusedField: fields[0],
				pages: newContext
			}
		);
	}

	/**
	 * @param {!Array} pages
	 * @private
	 */

	_handlePagesUpdated(pages) {
		this.setState(
			{
				pages
			}
		);
	}

	/**
	 * @param {!Array} pages
	 * @param {!Object} source
	 * @private
	 * @return {Object}
	 */

	_removeEmptyRow(pages, source) {
		const {pageIndex, rowIndex} = source;

		if (!FormSupport.hasFieldsRow(pages, pageIndex, rowIndex)) {
			pages = FormSupport.removeRow(pages, pageIndex, rowIndex);
		}

		return pages;
	}

	/**
	 * @param {!Array} pages
	 * @param {!Object} target
	 * @param {!Object} field
	 * @private
	 * @return {Object}
	 */

	_addRow(pages, target, fields) {
		const {pageIndex, rowIndex} = target;
		const newRow = FormSupport.implAddRow(12, fields);

		return FormSupport.addRow(pages, rowIndex, pageIndex, newRow);
	}

	/**
	 * @param {!Array} pages
	 * @param {!Object} target
	 * @param {!Object} field
	 * @private
	 * @return {Object}
	 */

	_setColumnFields(pages, target, fields) {
		const {columnIndex, pageIndex, rowIndex} = target;

		return FormSupport.setColumnFields(
			pages,
			pageIndex,
			rowIndex,
			columnIndex,
			fields
		);
	}

	render() {
		const {children, spritemap} = this.props;
		const {focusedField, mode, pages} = this.state;

		if (children.length) {
			const events = {
				deleteField: this._handleDeleteField.bind(this),
				duplicateField: this._handleDuplicatedField.bind(this),
				fieldAdded: this._handleFieldAdd.bind(this),
				fieldClicked: this._handleClickedField.bind(this),
				fieldEdited: this._handleFieldEdited.bind(this),
				fieldMoved: this._handleFieldMoved.bind(this),
				pagesUpdated: this._handlePagesUpdated.bind(this)
			};

			for (let index = 0; index < children.length; index++) {
				const child = children[index];

				Object.assign(
					child.props,
					{
						...this.otherProps(),
						events,
						focusedField,
						mode,
						pages,
						spritemap
					}
				);
			}
		}

		return children;
	}
}

export default LayoutProvider;