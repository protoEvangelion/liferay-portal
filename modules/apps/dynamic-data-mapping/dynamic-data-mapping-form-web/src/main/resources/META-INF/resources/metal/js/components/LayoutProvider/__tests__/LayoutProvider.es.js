import {JSXComponent} from 'metal-jsx';
import Context from './__mock__/mockContext.es';
import LayoutProvider from '../LayoutProvider.es';

let component;
let pages = null;
const spritemap = 'icons.svg';

class Child extends JSXComponent {
	render() {
		return <div />;
	}
}

class Parent extends JSXComponent {
	render() {
		return (
			<LayoutProvider
				pages={pages}
				ref="provider"
				spritemap={spritemap}
			>
				<Child ref="child" />
			</LayoutProvider>
		);
	}
}

describe(
	'LayoutProvider',
	() => {
		beforeEach(
			() => {
				pages = JSON.parse(JSON.stringify(Context));

				jest.useFakeTimers();
			}
		);

		afterEach(
			() => {
				if (component) {
					component.dispose();
				}

				pages = null;
			}
		);

		it(
			'should receive pages through PROPS and move to the internal state',
			() => {
				component = new LayoutProvider(
					{
						pages
					}
				);

				expect(component.state.pages).toEqual(pages);
			}
		);

		it(
			'should attach the events to the child component',
			() => {
				component = new Parent();

				const {provider} = component.refs;

				expect(provider.props.children[0].props.events).toMatchObject(
					{
						deleteField: expect.any(Function),
						fieldAdded: expect.any(Function),
						fieldClicked: expect.any(Function),
						fieldEdited: expect.any(Function),
						fieldMoved: expect.any(Function)
					}
				);
			}
		);

		it(
			'should pass to the child component the pages of the internal state',
			() => {
				component = new Parent();

				const {child, provider} = component.refs;

				expect(child.props.pages).toEqual(provider.state.pages);
			}
		);

		it(
			'should pass to the child component the focusedField',
			() => {
				component = new Parent();

				const {child, provider} = component.refs;

				const focusedField = {
					columnIndex: 0,
					pageIndex: 0,
					rowIndex: 0,
					type: 'radio'
				};

				provider.setState(
					{
						focusedField
					}
				);

				jest.runAllTimers();

				expect(child.props.focusedField).toEqual(provider.state.focusedField);
			}
		);

		it(
			'should pass to the child component the mode',
			() => {
				component = new Parent();

				const {child, provider} = component.refs;

				provider.setState(
					{
						mode: 'edit'
					}
				);

				jest.runAllTimers();

				expect(child.props.mode).toEqual(provider.state.mode);
			}
		);

		describe(
			'Field events',
			() => {
				describe(
					'fieldMoved',
					() => {
						it(
							'should listen to the fieldMoved event and move the field to the row in pages',
							() => {
								component = new Parent();

								const {child, provider} = component.refs;
								const mockEvent = {
									source: {
										columnIndex: 0,
										pageIndex: 0,
										rowIndex: 1
									},
									target: {
										columnIndex: false,
										pageIndex: 0,
										rowIndex: 0
									}
								};

								child.emit('fieldMoved', mockEvent);

								jest.runAllTimers();

								expect(provider.state.pages).toMatchSnapshot();
								expect(child.props.pages).toEqual(provider.state.pages);
							}
						);

						it(
							'should listen to the pagesUpdated event',
							() => {
								component = new Parent();

								const {child, provider} = component.refs;

								child.emit('pagesUpdated', pages);

								jest.runAllTimers();

								expect(provider.state.pages).toMatchSnapshot();
								expect(child.props.pages).toEqual(provider.state.pages);
							}
						);

						it(
							'should listen to the fieldMoved event and move the field to the column in pages',
							() => {
								component = new Parent();

								const {child, provider} = component.refs;
								const mockEvent = {
									source: {
										columnIndex: 1,
										pageIndex: 0,
										rowIndex: 1
									},
									target: {
										columnIndex: 1,
										pageIndex: 0,
										rowIndex: 0
									}
								};

								child.emit('fieldMoved', mockEvent);

								jest.runAllTimers();

								expect(provider.state.pages).toMatchSnapshot();
								expect(child.props.pages).toEqual(provider.state.pages);
							}
						);

						it(
							'should move the field to the column in pages and remove the row if there are no fields',
							() => {
								component = new Parent();

								const {child, provider} = component.refs;
								const mockEvent = {
									source: {
										columnIndex: 0,
										pageIndex: 0,
										rowIndex: 2
									},
									target: {
										columnIndex: 1,
										pageIndex: 0,
										rowIndex: 0
									}
								};

								child.emit('fieldMoved', mockEvent);

								jest.runAllTimers();

								expect(provider.state.pages).toMatchSnapshot();
								expect(child.props.pages).toEqual(provider.state.pages);
							}
						);

						it(
							'should move the field to the row in pages and remove the row if there are no fields',
							() => {
								component = new Parent();

								const {child, provider} = component.refs;
								const mockEvent = {
									source: {
										columnIndex: 0,
										pageIndex: 0,
										rowIndex: 0
									},
									target: {
										columnIndex: false,
										pageIndex: 0,
										rowIndex: 0
									}
								};

								child.emit('fieldMoved', mockEvent);

								jest.runAllTimers();

								expect(provider.state.pages).toMatchSnapshot();
								expect(child.props.pages).toEqual(provider.state.pages);
							}
						);
					}
				);

				describe(
					'fieldAdded',
					() => {
						it(
							'should listen the fieldAdded event and add the field in the column to the pages',
							() => {
								component = new Parent();

								const {child, provider} = component.refs;
								const mockEvent = {
									fieldProperties: {
										type: 'text'
									},
									target: {
										columnIndex: 1,
										pageIndex: 0,
										rowIndex: 0
									}
								};

								child.emit('fieldAdded', mockEvent);

								jest.runAllTimers();

								expect(provider.state.pages).toMatchSnapshot();
								expect(child.props.pages).toEqual(provider.state.pages);
							}
						);

						it(
							'should listen the fieldAdded event and add the field in the row to the pages',
							() => {
								component = new Parent();

								const {child, provider} = component.refs;
								const mockEvent = {
									fieldProperties: {
										type: 'text'
									},
									target: {
										columnIndex: false,
										pageIndex: 0,
										rowIndex: 0
									}
								};

								child.emit('fieldAdded', mockEvent);

								jest.runAllTimers();

								expect(provider.state.pages).toMatchSnapshot();
								expect(child.props.pages).toEqual(provider.state.pages);
							}
						);

						it(
							'should update the focusedField with the location of the new field when adding to the pages',
							() => {
								component = new Parent();

								const {child, provider} = component.refs;
								const mockEvent = {
									fieldProperties: {
										type: 'text'
									},
									target: {
										columnIndex: 2,
										pageIndex: 0,
										rowIndex: 1
									}
								};

								child.emit('fieldAdded', mockEvent);

								const focusedField = {
									...mockEvent.target,
									type: mockEvent.fieldProperties.type
								};

								jest.runAllTimers();

								expect(provider.state.focusedField).toEqual(focusedField);
								expect(child.props.focusedField).toEqual(focusedField);
								expect(provider.state.mode).toBe('edit');
							}
						);
					}
				);

				describe(
					'deleteField',
					() => {
						it(
							'should listen the fieldDeleted event and delete the field in the column to the pages',
							() => {
								component = new Parent();

								const {child, provider} = component.refs;
								const mockEvent = {
									columnIndex: 0,
									pageIndex: 0,
									rowIndex: 1
								};

								child.emit('deleteField', mockEvent);

								jest.runAllTimers();

								expect(provider.state.pages).toMatchSnapshot();
								expect(child.props.pages).toEqual(provider.state.pages);
							}
						);
					}
				);

				describe(
					'duplicateField',
					() => {
						it(
							'should listen the duplicate field event and add this field in the pages',
							() => {
								component = new Parent();

								const {child, provider} = component.refs;
								const mockEvent = {
									columnIndex: 0,
									pageIndex: 0,
									rowIndex: 0
								};

								child.emit('duplicateField', mockEvent);

								jest.runAllTimers();

								expect(provider.state.pages).toMatchSnapshot();
								expect(child.props.pages).toEqual(provider.state.pages);
							}
						);
					}
				);

				describe(
					'fieldEdited',
					() => {
						it(
							'should listen the fieldEdited event and edit the field to the pages',
							() => {
								component = new Parent();

								const {child, provider} = component.refs;
								const mockEvent = {
									key: 'label',
									value: 'Foo'
								};
								const mockFieldFocus = {
									columnIndex: 0,
									pageIndex: 0,
									rowIndex: 1
								};

								child.emit('fieldClicked', mockFieldFocus);
								child.emit('fieldEdited', mockEvent);

								jest.runAllTimers();

								expect(
									provider.state.pages[mockFieldFocus.pageIndex].rows[
										mockFieldFocus.rowIndex
									].columns[mockFieldFocus.columnIndex].fields[0][
										mockEvent.key
									]
								).toBe(mockEvent.value);
								expect(provider.state.pages).toMatchSnapshot();
								expect(child.props.pages).toEqual(provider.state.pages);
							}
						);
					}
				);

				describe(
					'fieldClicked',
					() => {
						it(
							'should listen the fieldClicked event and change the state of the focusedField to the data receive',
							() => {
								component = new Parent();

								const {child, provider} = component.refs;
								const mockEvent = {
									columnIndex: 0,
									mode: 'edit',
									pageIndex: 0,
									rowIndex: 0
								};

								child.emit('fieldClicked', mockEvent);

								expect(provider.props.children[0].props.events).toMatchObject(
									{
										fieldClicked: expect.any(Function)
									}
								);
								expect(provider.state.focusedField).toEqual(mockEvent);
							}
						);
					}
				);
			}
		);
	}
);