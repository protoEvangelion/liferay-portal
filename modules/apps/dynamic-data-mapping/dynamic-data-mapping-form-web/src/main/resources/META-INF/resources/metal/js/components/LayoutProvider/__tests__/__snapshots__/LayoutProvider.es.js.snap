// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoutProvider Field events deleteField should listen the fieldDeleted event and delete the field in the column to the pages 1`] = `
Array [
  Object {
    "rows": Array [
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Radio",
                "type": "radio",
              },
            ],
            "size": 3,
          },
          Object {
            "fields": Array [],
            "size": 9,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [],
            "size": 4,
          },
          Object {
            "fields": Array [
              Object {
                "items": Array [
                  Object {
                    "name": "Foo",
                  },
                  Object {
                    "name": "Bar",
                  },
                ],
                "label": "Select",
                "required": true,
                "type": "select",
              },
            ],
            "size": 6,
          },
          Object {
            "fields": Array [],
            "size": 2,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Second row",
                "required": true,
                "type": "text",
              },
            ],
            "size": 12,
          },
        ],
      },
    ],
    "title": "Untitled name",
  },
]
`;

exports[`LayoutProvider Field events duplicateField should listen the duplicate field event and add this field in the pages 1`] = `
Array [
  Object {
    "rows": Array [
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Radio",
                "type": "radio",
              },
            ],
            "size": 3,
          },
          Object {
            "fields": Array [],
            "size": 9,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "copy-of-x",
                "name": NaN,
                "type": "radio",
              },
            ],
            "size": 12,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Text",
                "required": true,
                "type": "text",
              },
            ],
            "size": 4,
          },
          Object {
            "fields": Array [
              Object {
                "items": Array [
                  Object {
                    "name": "Foo",
                  },
                  Object {
                    "name": "Bar",
                  },
                ],
                "label": "Select",
                "required": true,
                "type": "select",
              },
            ],
            "size": 6,
          },
          Object {
            "fields": Array [],
            "size": 2,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Second row",
                "required": true,
                "type": "text",
              },
            ],
            "size": 12,
          },
        ],
      },
    ],
    "title": "Untitled name",
  },
]
`;

exports[`LayoutProvider Field events fieldAdded should listen the fieldAdded event and add the field in the column to the pages 1`] = `
Array [
  Object {
    "rows": Array [
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Radio",
                "type": "radio",
              },
            ],
            "size": 3,
          },
          Object {
            "fields": Array [
              Object {
                "spritemap": "icons.svg",
                "type": "text",
              },
            ],
            "size": 9,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Text",
                "required": true,
                "type": "text",
              },
            ],
            "size": 4,
          },
          Object {
            "fields": Array [
              Object {
                "items": Array [
                  Object {
                    "name": "Foo",
                  },
                  Object {
                    "name": "Bar",
                  },
                ],
                "label": "Select",
                "required": true,
                "type": "select",
              },
            ],
            "size": 6,
          },
          Object {
            "fields": Array [],
            "size": 2,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Second row",
                "required": true,
                "type": "text",
              },
            ],
            "size": 12,
          },
        ],
      },
    ],
    "title": "Untitled name",
  },
]
`;

exports[`LayoutProvider Field events fieldAdded should listen the fieldAdded event and add the field in the row to the pages 1`] = `
Array [
  Object {
    "rows": Array [
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "spritemap": "icons.svg",
                "type": "text",
              },
            ],
            "size": 12,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Radio",
                "type": "radio",
              },
            ],
            "size": 3,
          },
          Object {
            "fields": Array [],
            "size": 9,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Text",
                "required": true,
                "type": "text",
              },
            ],
            "size": 4,
          },
          Object {
            "fields": Array [
              Object {
                "items": Array [
                  Object {
                    "name": "Foo",
                  },
                  Object {
                    "name": "Bar",
                  },
                ],
                "label": "Select",
                "required": true,
                "type": "select",
              },
            ],
            "size": 6,
          },
          Object {
            "fields": Array [],
            "size": 2,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Second row",
                "required": true,
                "type": "text",
              },
            ],
            "size": 12,
          },
        ],
      },
    ],
    "title": "Untitled name",
  },
]
`;

exports[`LayoutProvider Field events fieldEdited should listen the fieldEdited event and edit the field to the pages 1`] = `
Array [
  Object {
    "rows": Array [
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Radio",
                "type": "radio",
              },
            ],
            "size": 3,
          },
          Object {
            "fields": Array [],
            "size": 9,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Foo",
                "required": true,
                "type": "text",
              },
            ],
            "size": 4,
          },
          Object {
            "fields": Array [
              Object {
                "items": Array [
                  Object {
                    "name": "Foo",
                  },
                  Object {
                    "name": "Bar",
                  },
                ],
                "label": "Select",
                "required": true,
                "type": "select",
              },
            ],
            "size": 6,
          },
          Object {
            "fields": Array [],
            "size": 2,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Second row",
                "required": true,
                "type": "text",
              },
            ],
            "size": 12,
          },
        ],
      },
    ],
    "title": "Untitled name",
  },
]
`;

exports[`LayoutProvider Field events fieldMoved should listen to the fieldMoved event and move the field to the column in pages 1`] = `
Array [
  Object {
    "rows": Array [
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Radio",
                "type": "radio",
              },
            ],
            "size": 3,
          },
          Object {
            "fields": Array [
              Object {
                "items": Array [
                  Object {
                    "name": "Foo",
                  },
                  Object {
                    "name": "Bar",
                  },
                ],
                "label": "Select",
                "required": true,
                "type": "select",
              },
            ],
            "size": 9,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Text",
                "required": true,
                "type": "text",
              },
            ],
            "size": 4,
          },
          Object {
            "fields": Array [],
            "size": 6,
          },
          Object {
            "fields": Array [],
            "size": 2,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Second row",
                "required": true,
                "type": "text",
              },
            ],
            "size": 12,
          },
        ],
      },
    ],
    "title": "Untitled name",
  },
]
`;

exports[`LayoutProvider Field events fieldMoved should listen to the fieldMoved event and move the field to the row in pages 1`] = `
Array [
  Object {
    "rows": Array [
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Text",
                "required": true,
                "type": "text",
              },
            ],
            "size": 12,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Radio",
                "type": "radio",
              },
            ],
            "size": 3,
          },
          Object {
            "fields": Array [],
            "size": 9,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [],
            "size": 4,
          },
          Object {
            "fields": Array [
              Object {
                "items": Array [
                  Object {
                    "name": "Foo",
                  },
                  Object {
                    "name": "Bar",
                  },
                ],
                "label": "Select",
                "required": true,
                "type": "select",
              },
            ],
            "size": 6,
          },
          Object {
            "fields": Array [],
            "size": 2,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Second row",
                "required": true,
                "type": "text",
              },
            ],
            "size": 12,
          },
        ],
      },
    ],
    "title": "Untitled name",
  },
]
`;

exports[`LayoutProvider Field events fieldMoved should listen to the pagesUpdated event 1`] = `
Array [
  Object {
    "rows": Array [
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Radio",
                "type": "radio",
              },
            ],
            "size": 3,
          },
          Object {
            "fields": Array [],
            "size": 9,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Text",
                "required": true,
                "type": "text",
              },
            ],
            "size": 4,
          },
          Object {
            "fields": Array [
              Object {
                "items": Array [
                  Object {
                    "name": "Foo",
                  },
                  Object {
                    "name": "Bar",
                  },
                ],
                "label": "Select",
                "required": true,
                "type": "select",
              },
            ],
            "size": 6,
          },
          Object {
            "fields": Array [],
            "size": 2,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Second row",
                "required": true,
                "type": "text",
              },
            ],
            "size": 12,
          },
        ],
      },
    ],
    "title": "Untitled name",
  },
]
`;

exports[`LayoutProvider Field events fieldMoved should move the field to the column in pages and remove the row if there are no fields 1`] = `
Array [
  Object {
    "rows": Array [
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Radio",
                "type": "radio",
              },
            ],
            "size": 3,
          },
          Object {
            "fields": Array [
              Object {
                "label": "Second row",
                "required": true,
                "type": "text",
              },
            ],
            "size": 9,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Text",
                "required": true,
                "type": "text",
              },
            ],
            "size": 4,
          },
          Object {
            "fields": Array [
              Object {
                "items": Array [
                  Object {
                    "name": "Foo",
                  },
                  Object {
                    "name": "Bar",
                  },
                ],
                "label": "Select",
                "required": true,
                "type": "select",
              },
            ],
            "size": 6,
          },
          Object {
            "fields": Array [],
            "size": 2,
          },
        ],
      },
    ],
    "title": "Untitled name",
  },
]
`;

exports[`LayoutProvider Field events fieldMoved should move the field to the row in pages and remove the row if there are no fields 1`] = `
Array [
  Object {
    "rows": Array [
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Radio",
                "type": "radio",
              },
            ],
            "size": 12,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Text",
                "required": true,
                "type": "text",
              },
            ],
            "size": 4,
          },
          Object {
            "fields": Array [
              Object {
                "items": Array [
                  Object {
                    "name": "Foo",
                  },
                  Object {
                    "name": "Bar",
                  },
                ],
                "label": "Select",
                "required": true,
                "type": "select",
              },
            ],
            "size": 6,
          },
          Object {
            "fields": Array [],
            "size": 2,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Second row",
                "required": true,
                "type": "text",
              },
            ],
            "size": 12,
          },
        ],
      },
    ],
    "title": "Untitled name",
  },
]
`;
